### PR TITLE
Fix device handling and logits concatenation in OliveEvaluator

### PR DESCRIPTION
**Pull request by tezheng**
- Add exception handling for unsupported devices in `device_string_to_torch_device` method.
- Correct logits concatenation in `OnnxEvaluator` by using `logits_dict` instead of `logits`.
- Initialize `logits_dict` in `PyTorchEvaluator` to handle different result types.
- Update `_inference` method in `PyTorchEvaluator` to handle different result types and concatenate logits correctly.

## Checklist before requesting a review
- [x] Add unit tests for this change.
- [x] Make sure all tests can pass.
- [ ] Update documents if necessary.
- [x] Lint and apply fixes to your code by running `lintrunner -a`
- [ ] Is this a user-facing change? If yes, give a description of this change to be included in the release notes.
- [ ] Is this PR including examples changes? If yes, please remember to update [example documentation](https://github.com/microsoft/Olive/blob/main/docs/source/examples.md) in a follow-up PR.

### CODE REVIEW

File: olive/evaluator/olive_evaluator.py [hunk 1]
```diff
@@ -196,7 +196,11 @@ def compute_throughput(metric: Metric, latencies: Any) -> MetricResult:
 class _OliveEvaluator(OliveEvaluator):
     @staticmethod
     def device_string_to_torch_device(device: Device):
-        return torch.device("cuda") if device == Device.GPU else torch.device(device)
+        try:
```
Comment (jambayk) [inline olive/evaluator/olive_evaluator.py:199]: maybe we can just do `torch.device("cuda") if device == Device.GPU else torch.device("cpu")`. Otherwise, this would raise the warning for npu targets everytime. since this method is only used when evaluating pytorch models, I think it's save to just return cuda or cpu only.

Comment (tezheng) [reply olive/evaluator/olive_evaluator.py:199]: So we will only consider `cuda` and `cpu`, for now at least? Are we planning to support device like Apple `mps`?

Comment (jambayk) [reply olive/evaluator/olive_evaluator.py:199]: yes, there are no plans for other torch devices like mps for now.

File: olive/evaluator/olive_evaluator.py [hunk 2]
```diff
@@ -748,15 +753,20 @@ def _inference(
...
-                if not isinstance(result, torch.Tensor) and getattr(result, "logits", None) is not None
-                else result.cpu()
-            )
+            if isinstance(result, torch.Tensor):
+                logits.append(result.cpu())
+            elif isinstance(result, (list, tuple)):
+                logits.append([r.cpu() for r in result])
+            elif isinstance(result, dict):
+                for k in result:
+                    logits_dict[k].append(result[k].cpu())
```
Comment (xiaoyu-work) [inline olive/evaluator/olive_evaluator.py:762]: How can we handle the case when result has logits attribute?

Comment (tezheng) [reply olive/evaluator/olive_evaluator.py:762]: You mean result like `{ "logits": tensor }`?

**Comment by jambayk** [comment]
closing since it has become stale and dict issue was resolved in #701. Might still need the device solution.
